airflow/api/common/experimental/mark_tasks.py

```diff
@@ -104,6 +120,16 @@ def set_state(
     if dag is None:
         raise ValueError("Received tasks with no DAG")
 
+    if run_id is not None:
+        dr = (
+            session.query(DagRun.execution_date)
+            .filter(DagRun.run_id == run_id, DagRun.dag_id == dag.dag_id)
+            .one_or_none()
+        )
+        if dr is None:
+            raise ValueError(f"Run id {run_id} not found")
+        execution_date = dr.execution_date
+
     dates = get_execution_dates(dag, execution_date, future, past)
```

Comment:
Hmmm, if a run_id is given, we should really just use that to look up things, rather than convert it back to a date.

Reply:
It looks like if we don't convert it here, we will eventually convert to `execution_date ` in `get_execution_date` when we want to get the execution dates for the DAG since we will need to get the `start_date` and `end_date` to use in checking the execution dates if future or past is supplied. 
I can't really figure how to get the execution dates without converting the run_id to execution_date. 
Working on it but will appreciate any clue

Reply:
The real problem is actually `get_execution_date`. In stead of getting a list of dates and then look up task instances from them (`get_all_dag_task_query`), we should instead get a list of `run_id`s and looks up task instances of those instead. This entire function should be entirely reconstructed to operate around DAG runs instead of dates, and all of `get_execution_dates`, `verify_dag_run_integrity`, `get_subdag_runs`, and `get_all_dag_task_query` should either be deprecated or removed/rewritten entirely.

Reply:
I deprecated the functions inside the module and created another module where I used `run_id`. I had to use `deprecated` lib because I'm thinking maybe someone uses the API externally. Let me know what you think about the whole thing.
Thanks!